This is the same patch as in nixpkgs but adapted to gdb-nuclei.
https://github.com/NixOS/nixpkgs/blob/e7fe577d9803885d1191c6612b95c246cb605dde/pkgs/development/tools/misc/gdb/debug-info-from-env.patch

--- a/gdb/main.c	2021-01-24 06:47:30.946031294 +0100
+++ b/gdb/main.c	2021-01-24 06:49:59.989105428 +0100
@@ -553,9 +553,17 @@
       gdb_sysroot = xstrdup (TARGET_SYSROOT_PREFIX);
     }
 
-  debug_file_directory = relocate_gdb_directory (DEBUGDIR,
-						 DEBUGDIR_RELOCATABLE);
-
+  debug_file_directory = getenv ("NIX_DEBUG_INFO_DIRS");
+  if (debug_file_directory != NULL)
+    // This might be updated later using
+    // $ set debug-file-directory /to/some/path
+    // which will use xfree. We must then have a xmallocated
+    // copy of the string that can be xfeed later.
+    debug_file_directory = xstrdup (debug_file_directory);
+  else
+    debug_file_directory = relocate_gdb_directory (DEBUGDIR,
+                                                   DEBUGDIR_RELOCATABLE);
+ 
   gdb_datadir = relocate_gdb_directory (GDB_DATADIR,
 					GDB_DATADIR_RELOCATABLE);
 
